File: samples/Haskell/probabilistic-choice.hs
import System.Random
import Data.List
import Control.Monad
import Control.Arrow

labels = ["aleph", "beth", "gimel", "daleth", "he", "waw", "zayin", "heth" ]
piv n = take n . (++  repeat ' ')

main = do
  g <- newStdGen
  let rs,ps :: [Float]
      rs = take 1000000 $ randomRs(0,1) g
      ps = ap (++) (return. (1 -) .sum) $ map recip [5..11]
      sps = scanl1 (+) ps
      qs = (\xs -> map ((/1000000.0).fromIntegral.length. flip filter xs. (==))sps)
           $ map (head . flip dropWhile sps . (>)) rs
  putStrLn $ "       expected     actual"
  mapM_ putStrLn $ zipWith3
               (\l s c-> (piv 7 l) ++ (piv 13 $ show $ s)
                    ++(piv 12 $ show $ c)) labels ps qs

File: samples/OCaml/probabilistic-choice.ml
let p = [
    "Aleph",   1.0 /. 5.0;
    "Beth",    1.0 /. 6.0;
    "Gimel",   1.0 /. 7.0;
    "Daleth",  1.0 /. 8.0;
    "He",      1.0 /. 9.0;
    "Waw",     1.0 /. 10.0;
    "Zayin",   1.0 /. 11.0;
    "Heth", 1759.0 /. 27720.0;
  ]

let rec take k = function
  | (v, p)::tl -> if k < p then v else take (k -. p) tl
  | _ -> invalid_arg "take"

let () =
  let n = 1_000_000 in
  Random.self_init();
  let h = Hashtbl.create 3 in
  List.iter (fun (v, _) -> Hashtbl.add h v 0) p;
  let tot = List.fold_left (fun acc (_, p) -> acc +. p) 0.0 p in
  for i = 1 to n do
    let sel = take (Random.float tot) p in
    let n = Hashtbl.find h sel in
    Hashtbl.replace h sel (succ n)  (* count the number of each item *)
  done;
  List.iter (fun (v, p) ->
    let d = Hashtbl.find h v in
    Printf.printf "%s \t %f %f\n" v p (float d /. float n)
  ) p
